File: PLASMA/trunk/src/PLASMA/ConstraintEngine/component/DataTypes.cc
/*
 * DataTypes.cc
 *
 *  Created on: Mar 31, 2009
 *      Author: javier
 */

#include "DataTypes.hh"

#include "BoolDomain.hh"
#include "EnumeratedDomain.hh"
#include "IntervalDomain.hh"
#include "IntervalIntDomain.hh"
#include "StringDomain.hh"
#include "SymbolDomain.hh"
#include "Variable.hh"

#define DT_STATIC_MEMBERS(dataType,dtName) \
const std::string& dataType::NAME() { static std::string sl_name(#dtName); return sl_name; } \
dataType& dataType::instance() { static dataType sl_instance; return sl_instance; }


namespace EUROPA
{

DT_STATIC_MEMBERS(VoidDT,void)
DT_STATIC_MEMBERS(FloatDT,float)
DT_STATIC_MEMBERS(IntDT,int)
DT_STATIC_MEMBERS(BoolDT,bool)
DT_STATIC_MEMBERS(StringDT,string)
DT_STATIC_MEMBERS(SymbolDT,symbol)


VoidDT::VoidDT()
    : DataType(NAME().c_str())
{
}

VoidDT::~VoidDT()
{
}

bool VoidDT::isNumeric() const { return false; }
bool VoidDT::isBool() const  { return false; }
bool VoidDT::isString() const  { return false; }

double VoidDT::createValue(const std::string& value) const
{
  check_error(ALWAYS_FAILS, "can't create void value");
  return -1;
}

ConstrainedVariableId
VoidDT::createVariable(const ConstraintEngineId& constraintEngine,
                                const AbstractDomain& baseDomain,
                                const bool internal,
                                bool canBeSpecified,
                                const char* name,
                                const EntityId& parent,
                                int index) const
{
    check_error(ALWAYS_FAILS, "can't create void variable");
    return ConstrainedVariableId::noId();
}

FloatDT::FloatDT()
    : DataType(NAME().c_str())
{
    m_baseDomain = new IntervalDomain(NAME().c_str());
}

FloatDT::~FloatDT()
{
}

bool FloatDT::isNumeric() const { return true; }
bool FloatDT::isBool() const  { return false; }
bool FloatDT::isString() const  { return false; }

double FloatDT::createValue(const std::string& value) const
{
  // TODO: simplify this
  if (value == "-inf" || value=="-inff") {
    return MINUS_INFINITY;
  }
  if (value == "inf" || value=="+inf" || value=="inff" || value=="+inff") {
    return PLUS_INFINITY;
  }
  return atof(value.c_str());
}

IntDT::IntDT()
    : DataType(NAME().c_str())
{
    m_baseDomain = new IntervalIntDomain(NAME().c_str());
}

IntDT::~IntDT()
{
}

bool IntDT::isNumeric() const { return true; }
bool IntDT::isBool() const  { return false; }
bool IntDT::isString() const  { return false; }

double IntDT::createValue(const std::string& value) const
{
  if (value == "-inf") {
    return MINUS_INFINITY;
  }
  if (value == "+inf" || value == "inf") {
    return PLUS_INFINITY;
  }
  return atoi(value.c_str());
}

BoolDT::BoolDT()
    : DataType(NAME().c_str())
{
    m_baseDomain = new BoolDomain(NAME().c_str());
}

BoolDT::~BoolDT()
{
}

/*
 * To maintain compatibility with the original definition of the Bool domain in EUROPA :
 * BoolDT represents the integer interval [0,1]
 */
bool BoolDT::isNumeric() const { return true; }
bool BoolDT::isBool() const  { return true; }
bool BoolDT::isString() const  { return false; }

double BoolDT::createValue(const std::string& value) const
{
  if (value == "true") {
    return true;
  }
  if (value == "false") {
    return false;
  }
  check_error(ALWAYS_FAILS, "string value for boolean should be 'true' or 'false', not '" + value + "'");
  return -1;
}

StringDT::StringDT()
    : DataType(NAME().c_str())
{
    m_baseDomain = new StringDomain(NAME().c_str());
}

StringDT::~StringDT()
{
}

bool StringDT::isNumeric() const { return false; }
bool StringDT::isBool() const  { return false; }
bool StringDT::isString() const  { return true; }

double StringDT::createValue(const std::string& value) const
{
  return LabelStr(value);
}

SymbolDT::SymbolDT()
    : DataType(NAME().c_str())
{
    m_baseDomain = new StringDomain(NAME().c_str());
}

SymbolDT::~SymbolDT()
{
}

bool SymbolDT::isNumeric() const { return false; }
bool SymbolDT::isBool() const  { return false; }
bool SymbolDT::isString() const  { return true; }

double SymbolDT::createValue(const std::string& value) const
{
  return LabelStr(value);
}

ObjectDT::ObjectDT(const char* name, const AbstractDomain& baseDomain)
    : DataType(name)
{
    m_baseDomain = baseDomain.copy();
}

ObjectDT::~ObjectDT()
{
}

bool ObjectDT::isNumeric() const { return false; }
bool ObjectDT::isBool() const  { return false; }
bool ObjectDT::isString() const  { return false; }
bool ObjectDT::isEntity() const  { return true; }

double ObjectDT::createValue(const std::string& value) const
{
  return LabelStr(value);
}

RestrictedDT::RestrictedDT(const char* name, const DataTypeId& baseType, const AbstractDomain& baseDomain)
    : DataType(name)
    , m_baseType(baseType)
{
    m_baseDomain = baseDomain.copy();
    setIsRestricted(true);
}

RestrictedDT::~RestrictedDT()
{
}

bool RestrictedDT::isNumeric() const { return m_baseType->isNumeric(); }
bool RestrictedDT::isBool() const  { return m_baseType->isBool(); }
bool RestrictedDT::isString() const  { return m_baseType->isString(); }
bool RestrictedDT::isEntity() const  { return m_baseType->isEntity(); }

double RestrictedDT::createValue(const std::string& value) const
{
    if (isNumeric())
        return(atof(value.c_str()));

    return(LabelStr(value));
}


}

File: Incubation/trunk/scala-europa/src/main/scala/ConstraintEngine/component/DataTypes.scala
package gov.nasa.arc.europa.constraintengine.component
import gov.nasa.arc.europa.constraintengine.DataType
import gov.nasa.arc.europa.constraintengine.Domain
import gov.nasa.arc.europa.utils.Number._

class VoidDT extends DataType(VoidDT.NAME) { 
  override val baseDomain = Domain.NO_DOMAIN
  override def isNumeric = false
  override def isBool = false
  override def isString = false
  override def createValue(v: String): Option[Double] = throw new Exception("Can't create an instance of the void type.")
  override def emptyDomain = Domain.NO_DOMAIN
}
object VoidDT { 
  val NAME = "void"
  val INSTANCE = new VoidDT
}

class FloatDT extends DataType(FloatDT.NAME, 0.00001, null) { 
  override val baseDomain = new IntervalDomain(this)
  override def isNumeric = true
  override def isBool = false
  override def isString = false
  override def createValue(v: String): Option[Double] = { 
    if(v == "-inf" || v == "-inff") return Some(MINUS_INFINITY)
    else if(v == "inf" || v == "+inf" || v == "inff" || v == "+inff") return Some(PLUS_INFINITY)
    else return try { Some(java.lang.Double.parseDouble(v))} catch {case _ => None}
  }
  override def emptyDomain = new IntervalDomain(3, -2)
}

object FloatDT { 
  val NAME = "float"
  val INSTANCE = new FloatDT
}

class IntDT extends DataType(IntDT.NAME, 1, null) { 
  override val baseDomain = new IntervalIntDomain()
  override def isNumeric = true
  override def isBool = false
  override def isString = false
  override def createValue(v: String): Option[Double] = { 
    if(v == "-inf") return Some(MINUS_INFINITY)
    else if(v == "inf" || v == "+inf") return Some(PLUS_INFINITY)
    else return try { Some(java.lang.Double.parseDouble(v))} catch {case _ => None}
  }
  override def emptyDomain = new IntervalIntDomain(3, -2)
  override def toString(value: Double): String = value.toInt.toString
}

object IntDT { 
  val NAME = "int"
  val INSTANCE = new IntDT
}

class BoolDT extends DataType(BoolDT.NAME, 1, null) { 
  override val baseDomain = new BoolDomain()
  override def isNumeric = true
  override def isBool = true
  override def isString = false
  override def createValue(v: String): Option[Double] = v match { 
    case "true" => Some(1)
    case "false" => Some(0)
    case _ => None
  }
  override def toString(v: Double): String = { 
    if(v == 0) "false"
    else "true"
  }
  override def emptyDomain = new BoolDomain(1.0, 0.0)
}

object BoolDT { 
  val NAME = "bool"
  val INSTANCE = new BoolDT
}
